src/test/java/uk/gov/ch/oauth/OAuthCoordinatorTest.java

```diff
@@ -0,0 +1,166 @@
+package uk.gov.ch.oauth;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import javax.servlet.http.HttpServletResponse;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.Spy;
+import org.mockito.junit.jupiter.MockitoExtension;
+import uk.gov.ch.oauth.exceptions.UnauthorisedException;
+import uk.gov.ch.oauth.identity.IIdentityProvider;
+import uk.gov.ch.oauth.session.SessionFactory;
+import uk.gov.companieshouse.environment.EnvironmentReader;
+
+
+@ExtendWith(MockitoExtension.class)
+class OAuthCoordinatorTest {
+
+    @Spy
+    @InjectMocks
+    OAuthCoordinator oAuthCoordinator;
+
+    @Mock
+    private HttpServletResponse mockResponse;
+    @Mock
+    private IIdentityProvider mockIdentityProvider;
+    @Mock
+    private IOauth mockOAuth;
+
+    @Nested
+    class CallbackTests {
```

Comment:
Why have you decided to nest these unit tests

Reply:
Because its a separation of tests of the utility methods and the actual functional code.
Additionally with three extra methods like Call back that will need to be added this class could get long and complicated so this adds a logical structure for someone reading it.

Reply:
Sure 👍